lib/promise.ml

```diff
@@ -0,0 +1,33 @@
+open Effect
+open Sched
+
+type 'a state = Full of 'a | Empty of 'a resumer list
+type 'a t = 'a state Atomic.t
+
+let create () = Atomic.make (Empty [])
+
+let take p =
+  let rec block r =
+    let old = Atomic.get p in
+    match old with
+    | Full v -> Some v
+    | Empty l ->
+      if Atomic.compare_and_set p old (Empty (r::l))
+      then None else block r
```

Comment:
A performance issue here is that this lock-free retry loop does not include [a backoff mechanism to reduce contention](https://github.com/ocaml-multicore/backoff#about-contention).  There is also technically no guarantee for lack of starvation with this kind of retry loop in general &mdash; it could be that other CPU cores repeatedly manage to modify the target of the CAS.

You could, of course, add a backoff here, but I'd like to raise the question of whether it is a good idea to have such non-starvation free algorithms in this context.  The issue is that this retry loop is being executed inside the effect handler, which is part of the scheduler (not part of the user code).  During the time the retry loop is being run, it could be that the fiber is cancelled (on a scheduler that allows cross domain cancellation), for example.  What are the semantics in such cases?  What considerations does this impose on the design and implementation of schedulers?  Can and do `Suspend` handlers of schedulers need to protect against misbehaving `block`s (that might never return or might raise exceptions or perform effects) passed to them via `Suspend` effects?

To me the idea that OCaml would standardize on a `Suspend` effect interface that allows arbitrary code to be able to pass arbitrary closures to be run inside critical sections of schedulers seems rather troublesome.  This is a major benefit of the [DLA interface](https://github.com/ocaml-multicore/domain-local-await/#readme), which does not introduce such a backdoor.  To cast the DLA interface as an effect, one could introduce an effect such as:

```ocaml
type single_use_trigger = { release: unit -> unit; await : unit -> unit }

type _ Effect.t += Prepare_for_await : single_use_trigger Effect.t
```

Then, instead of calling `Domain_local_await.prepare_for_await ()`, one would `perform Prepare_for_await` to obtain a `single_use_trigger` and avoid having the scheduler run arbitrary user closures during effect handlers.